features/FxA-45-device-registration-api/FX-DESKTOP.md

```diff
@@ -0,0 +1,122 @@
+Firefox Device Registration
+===========================
+
+This is a breakdown of the states Firefox needs to handle/worry about.
+
+See the [README.md](README.md) file for a list of requirements and user
+stories.
+
+## New profile, no FxA deviceID, no FxA sessionToken
+
+Users of a new Firefox profile who have never signed up/in to an FxA backed
+service will have neither an FxA sessionToken nor an FxA device ID.
+
+Firefox should take no action.
+
+## Existing profile, no FxA deviceID, has FxA sessionToken
+
+This is a user who has previously signed in to FxA to connect
+to Sync. This user will have an FxA sessionToken, but may not yet
+have an FxA deviceID.
+
+The user facing goal is for the device to appear in the user's device
+list without additional user action. To facilitate this, Firefox
+should register the device on behalf of the user.
+
+On startup, Firefox should attempt to register the device with the
+FxA auth server using the name and type currently stored in the
+user's Sync preferences.
+
+Firefox will send the following information to the FxA auth server to
+register the device:
+
+* deviceName (from Sync preferences)
+* deviceType (from Sync preferences)
+* pushCallback (from Push preferences)
+* pushPublicKey (from Push preferences)
+
+### Registration success
+
+The FxA auth server will respond with:
+
+* deviceID
+* deviceName
+* deviceType
+* pushCallback
+* pushPublicKey
+
+Firefox must then store the returned deviceID.
+
+### Registration failure
+
+If registration fails because of a network error, Firefox should re-attempt
+to register after a TBD interval has elapsed.
+
+If registration fails for other reasons - WHAT HERE?
+
+## Existing profile, has FxA deviceID, has FxA sessionToken
+
+This is a user who has previously signed in to FxA to
+connect to Sync and their device is already registered.
+
+This device has already sent its initial name to the FxA
+auth server, but the user may have remotely disconnected
+the device, remotely reset the account password, or remotely
+changed the account password. The FxA auth server is considered
+the canonical source of truth for device state.
```

Comment:
You'll eventually be able to edit the device-name in both places, so I'm not sure "server-canonical" is the right way to think about it.  Would this mean that e.g. you can't edit the device-name in chrome UI when it's offline and can't save it to the server?